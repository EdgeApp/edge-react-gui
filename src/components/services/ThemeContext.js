// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
