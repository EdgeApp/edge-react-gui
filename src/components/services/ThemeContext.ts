import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
