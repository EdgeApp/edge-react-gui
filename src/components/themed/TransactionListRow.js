// @flow

import { abs, div, log10 } from 'biggystring'
import type { EdgeCurrencyInfo, EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'

import { getSymbolFromCurrency } from '../../constants/WalletAndCurrencyConstants.js'
import { displayFiatAmount } from '../../hooks/useFiatText.js'
import { formatNumber } from '../../locales/intl.js'
import s from '../../locales/strings'
import { getDisplayDenomination, getExchangeDenomination } from '../../selectors/DenominationSelectors.js'
import { connect } from '../../types/reactRedux.js'
import { type NavigationProp } from '../../types/routerTypes.js'
import type { TransactionListTx } from '../../types/types.js'
import {
  DECIMAL_PRECISION,
  decimalOrZero,
  DEFAULT_TRUNCATE_PRECISION,
  getDenomFromIsoCode,
  isSentTransaction,
  maxPrimaryCurrencyConversionDecimals,
  normalizeForSearch,
  precisionAdjust,
  truncateDecimals
} from '../../util/utils'
import { showError } from '../services/AirshipInstance.js'
import { TransactionRow } from './TransactionRow.js'

type StateProps = {
  cryptoAmount: string,
  denominationSymbol?: string,
  fiatAmount: string,
  fiatSymbol: string,
  isSentTransaction: boolean,
  requiredConfirmations: number,
  selectedCurrencyName: string,
  thumbnailPath?: string
}

type OwnProps = {
  // eslint-disable-next-line react/no-unused-prop-types
  walletId: string,
  // eslint-disable-next-line react/no-unused-prop-types
  currencyCode: string,
  transaction: TransactionListTx,
  navigation: NavigationProp<any>
}

type Props = OwnProps & StateProps

export class TransactionListRowComponent extends React.PureComponent<Props> {
  handlePress = () => {
    const { transaction, thumbnailPath, navigation } = this.props
    if (transaction == null) {
      return showError(s.strings.transaction_details_error_invalid)
    }
    navigation.push('transactionDetails', {
      edgeTransaction: transaction,
      thumbnailPath
    })
  }

  render() {
    return (
      <TransactionRow
        cryptoAmount={this.props.cryptoAmount}
        denominationSymbol={this.props.denominationSymbol}
        fiatAmount={this.props.fiatAmount}
        fiatSymbol={this.props.fiatSymbol}
        onPress={this.handlePress}
        isSentTransaction={this.props.isSentTransaction}
        requiredConfirmations={this.props.requiredConfirmations}
        selectedCurrencyName={this.props.selectedCurrencyName}
        thumbnailPath={this.props.thumbnailPath}
        transaction={this.props.transaction}
      />
    )
  }
}

export const TransactionListRow = connect<StateProps, {}, OwnProps>(
  (state, ownProps) => {
    const { currencyCode, walletId, transaction } = ownProps
    const { metadata } = transaction
    const { name, amountFiat } = metadata ?? {}
    const guiWallet = state.ui.wallets.byId[walletId]
    const { fiatCurrencyCode } = guiWallet
    const { currencyWallets } = state.core.account
    const coreWallet: EdgeCurrencyWallet = currencyWallets[walletId]
    const currencyInfo: EdgeCurrencyInfo = coreWallet.currencyInfo
    const displayDenomination = getDisplayDenomination(state, currencyInfo.pluginId, currencyCode)
    const exchangeDenomination = getExchangeDenomination(state, currencyInfo.pluginId, currencyCode)
    const fiatDenomination = getDenomFromIsoCode(guiWallet.fiatCurrencyCode)

    // Required Confirmations
    const requiredConfirmations = currencyInfo.requiredConfirmations || 1 // set default requiredConfirmations to 1, so once the transaction is in a block consider fully confirmed

    // Thumbnail
    let thumbnailPath
    const contacts = state.contacts || []
    const transactionContactName = name != null ? normalizeForSearch(name) : null
    for (const contact of contacts) {
      const { givenName, familyName } = contact
      const fullName = normalizeForSearch(`${givenName}${familyName ?? ''}`)
      if (contact.thumbnailPath && fullName === transactionContactName) {
        thumbnailPath = contact.thumbnailPath
        break
      }
    }

    // CryptoAmount
    const rateKey = `${currencyCode}_${guiWallet.isoFiatCurrencyCode}`
    const exchangeRate = state.exchangeRates[rateKey] ? state.exchangeRates[rateKey] : undefined
    let maxConversionDecimals = DEFAULT_TRUNCATE_PRECISION
    if (exchangeRate) {
      const precisionAdjustValue = precisionAdjust({
        primaryExchangeMultiplier: exchangeDenomination.multiplier,
        secondaryExchangeMultiplier: fiatDenomination.multiplier,
        exchangeSecondaryToPrimaryRatio: exchangeRate
      })
      maxConversionDecimals = maxPrimaryCurrencyConversionDecimals(log10(displayDenomination.multiplier), precisionAdjustValue)
    }
    const cryptoAmount = div(abs(transaction.nativeAmount ?? '0'), displayDenomination.multiplier, DECIMAL_PRECISION)
    const cryptoAmountFormat = formatNumber(decimalOrZero(truncateDecimals(cryptoAmount, maxConversionDecimals), maxConversionDecimals))

    return {
      isSentTransaction: isSentTransaction(transaction),
      cryptoAmount: cryptoAmountFormat,
      fiatAmount: displayFiatAmount(amountFiat),
      fiatSymbol: getSymbolFromCurrency(fiatCurrencyCode),
      denominationSymbol: displayDenomination.symbol,
      requiredConfirmations,
      selectedCurrencyName: guiWallet.currencyNames[currencyCode] || currencyCode,
      thumbnailPath
    }
  },
  dispatch => ({})
)(TransactionListRowComponent)
