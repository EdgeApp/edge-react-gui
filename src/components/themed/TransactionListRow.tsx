import { abs, div, gt, log10 } from 'biggystring'
import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import Share from 'react-native-share'
import { sprintf } from 'sprintf-js'

import { getSymbolFromCurrency } from '../../constants/WalletAndCurrencyConstants'
import { useContactThumbnail } from '../../hooks/redux/useContactThumbnail'
import { displayFiatAmount } from '../../hooks/useFiatText'
import { useHandler } from '../../hooks/useHandler'
import { useHistoricalRate } from '../../hooks/useHistoricalRate'
import { useWatch } from '../../hooks/useWatch'
import { formatNumber } from '../../locales/intl'
import s from '../../locales/strings'
import { getDisplayDenomination, getExchangeDenomination } from '../../selectors/DenominationSelectors'
import { useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { TransactionListTx } from '../../types/types'
import {
  DECIMAL_PRECISION,
  decimalOrZero,
  DEFAULT_TRUNCATE_PRECISION,
  getDenomFromIsoCode,
  isSentTransaction,
  maxPrimaryCurrencyConversionDecimals,
  precisionAdjust,
  truncateDecimals
} from '../../util/utils'
import { showError } from '../services/AirshipInstance'
import { TransactionRow } from './TransactionRow'

interface Props {
  navigation: NavigationBase
  wallet: EdgeCurrencyWallet
  currencyCode: string
  transaction: TransactionListTx
}

export function TransactionListRow(props: Props) {
  const { navigation, currencyCode, wallet, transaction } = props
  const { metadata } = transaction
  const { name, amountFiat: defaultAmountFiat = 0 } = metadata ?? {}

  const fiatCurrencyCode = useWatch(wallet, 'fiatCurrencyCode')
  const nonIsoFiatCurrencyCode = fiatCurrencyCode.replace('iso:', '')
  const currencyInfo = wallet.currencyInfo

  const displayDenomination = useSelector(state => getDisplayDenomination(state, currencyInfo.pluginId, currencyCode))
  const exchangeDenomination = useSelector(state => getExchangeDenomination(state, currencyInfo.pluginId, currencyCode))
  const fiatDenomination = getDenomFromIsoCode(nonIsoFiatCurrencyCode)

  const currencyName =
    currencyCode === currencyInfo.currencyCode
      ? currencyInfo.displayName
      : currencyInfo.metaTokens.find(metaToken => metaToken.currencyCode === currencyCode)?.currencyName

  // Required Confirmations
  const requiredConfirmations = currencyInfo.requiredConfirmations || 1 // set default requiredConfirmations to 1, so once the transaction is in a block consider fully confirmed

  // Thumbnail
  const thumbnailPath = useContactThumbnail(name)

  // CryptoAmount
  const rateKey = `${currencyCode}_${fiatCurrencyCode}`
  const exchangeRate: string = useSelector(state => state.exchangeRates[rateKey])
  let maxConversionDecimals = DEFAULT_TRUNCATE_PRECISION
  if (exchangeRate != null && gt(exchangeRate, '0')) {
    const precisionAdjustValue = precisionAdjust({
      primaryExchangeMultiplier: exchangeDenomination.multiplier,
      secondaryExchangeMultiplier: fiatDenomination.multiplier,
      exchangeSecondaryToPrimaryRatio: exchangeRate
    })
    maxConversionDecimals = maxPrimaryCurrencyConversionDecimals(log10(displayDenomination.multiplier), precisionAdjustValue)
  }
  const cryptoAmount = div(abs(transaction.nativeAmount ?? '0'), displayDenomination.multiplier, DECIMAL_PRECISION)
  const cryptoExchangeAmount = div(abs(transaction.nativeAmount ?? '0'), exchangeDenomination.multiplier, DECIMAL_PRECISION)
  const cryptoAmountFormat = formatNumber(decimalOrZero(truncateDecimals(cryptoAmount, maxConversionDecimals), maxConversionDecimals))

  const isoDate = new Date(transaction.date * 1000).toISOString()
  const historicalRate = useHistoricalRate(`${currencyCode}_${fiatCurrencyCode}`, isoDate)
  const amountFiat = defaultAmountFiat > 0 ? defaultAmountFiat : historicalRate * Number(cryptoExchangeAmount)

  const handlePress = useHandler(() => {
    if (transaction == null) {
      return showError(s.strings.transaction_details_error_invalid)
    }
    navigation.push('transactionDetails', {
      edgeTransaction: transaction,
      walletId: wallet.id
    })
  })

  const handleLongPress = useHandler(() => {
    const url = sprintf(currencyInfo.transactionExplorer, transaction.txid)
    const shareOptions = {
      url
    }
    Share.open(shareOptions).catch(e => showError(e))
  })

  return (
    <TransactionRow
      cryptoAmount={cryptoAmountFormat}
      denominationSymbol={displayDenomination.symbol}
      fiatAmount={displayFiatAmount(amountFiat)}
      fiatSymbol={getSymbolFromCurrency(nonIsoFiatCurrencyCode)}
      onPress={handlePress}
      onLongPress={handleLongPress}
      isSentTransaction={isSentTransaction(transaction)}
      requiredConfirmations={requiredConfirmations}
      selectedCurrencyName={currencyName || currencyCode}
      thumbnailPath={thumbnailPath}
      transaction={transaction}
      wallet={wallet}
    />
  )
}
