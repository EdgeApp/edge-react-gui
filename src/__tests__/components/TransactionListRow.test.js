/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext.js'
import { TransactionListRowComponent } from '../../components/themed/TransactionListRow.js'

describe('TransactionListRow', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      cryptoAmount: '1',
      denominationSymbol: 'BTC',
      fiatAmount: '1000000',
      fiatSymbol: 'USD',
      isSentTransaction: true,
      requiredConfirmations: 11,
      selectedCurrencyName: 'Bitcoin',
      thumbnailPath: 'wyre.png',
      walletBlockHeight: 11,
      walletId: '3dfgbA347...',
      currencyCode: 'BTC',
      transaction: '12ser4hh...',
      theme: getTheme()
    }
    const actual = renderer.render(<TransactionListRowComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
