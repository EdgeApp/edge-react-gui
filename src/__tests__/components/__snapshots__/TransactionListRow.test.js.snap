// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionListRow should render with loading props 1`] = `
<WithTheme(TransactionRowComponent)
  cryptoAmount="1"
  denominationSymbol="BTC"
  fiatAmount="1000000"
  fiatSymbol="USD"
  isSentTransaction={true}
  onPress={[Function]}
  requiredConfirmations={11}
  selectedCurrencyName="Bitcoin"
  thumbnailPath="wyre.png"
  transaction="12ser4hh..."
/>
`;
